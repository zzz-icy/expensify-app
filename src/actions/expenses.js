// Actions for expenses
// import uuid from 'uuid';
import database from '../firebase/firebase';
// action generators
// ADD-EXPENSE

// export const addExpense = (
//     { description = '', note = '', amount = 0, createdAt = 0 } = {}
// ) => (
//         {
//             type: 'ADD_EXPENSE',
//             expense: {
//                 id: uuid(),
//                 description,
//                 note,
//                 amount,
//                 createdAt,
//             }
//         });

// we do not need the id generated by uuid(), will use the id generated by firebase, so rewrite the addExpense

export const addExpense = (expense) => (
    {
        type: 'ADD_EXPENSE',
        expense
    });


export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        // add return so that we can chain .then() in the test
        return database.ref('expenses').push(expense).then((ref) => { //.then() gets called with the ref, so we have access to the id 
            dispatch(addExpense(
                {
                    id: ref.key,
                    ...expense
                }
            ));
        });
    }
};


// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id }) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            // once it is removed from database, dispatch action
            dispatch(removeExpense({ id }));
        });
    }
}
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// 1. fetch all expenses data
// 2. Parse that data into an array, cause data from firebase are nested objects with the id as the key for each expense
// 3. dispatch SET_EXPENSES

export const startSetExpenses = () => {
    return (dispatch) => {

        // add return so that we can chain .then() in the app.js
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expenses));
        });
    }
};