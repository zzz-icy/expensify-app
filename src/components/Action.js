import React from 'react';

// const Action = (props) => {
//     return (
//         <div>
//             <button
//                 disabled={!props.hasOption} // if class based should be this.props
//                 onClick={props.handlePick}
//             >
//                 What should I do?
//             </button>
//         </div>
//     );
// };

// refactored to use consice syntax
const Action = (props) => (
    <div>
        <button
            disabled={!props.hasOption} // if class based should be this.props
            onClick={props.handlePick}
        >
            What should I do?
            </button>
    </div>
);
;

export default Action;


// class Action extends React.Component {
//     // define method
//     // handlePick() {  // this is a method for Action component in new and consice syntax, outside render
//     //     alert('handlepick');
//     // }
//     render() {

//         return (
//             <div>
//                 <button disabled={!this.props.hasOption} onClick={this.props.handlePick} >What should I do?</button>
//             </div>
//         );
//     }
// }