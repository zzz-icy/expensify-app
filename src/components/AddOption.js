import React from 'react';
// import ReactDOM from 'react-dom';

export default class AddOption extends React.Component {
    state = {
        error: undefined,
    }
    // constructor(props) {
    //     super(props);
    //     this.handleAddOption = this.handleAddOption.bind(this);
    // the handleAddOption here refers to the method defined within AddOption component, not the props passed in
    // this.state = {
    //     error: undefined,
    // }
    // }

    handleAddOption = (e) => { // make it a arrow function, because arrow function has no data binding 
        e.preventDefault(); // this will stop the full page refreshing
        // console.log(e);
        const option = e.target.elements.option.value.trim();
        // .trim() take off spaces not internal
        // e.target is gonna point to the element the event started on
        // if (option) { // nolonger need if condition here, it's handled up above in the parent component
        // 
        // this.setState((prevState) => {
        //     return {
        //         options: prevState.options.push(option),
        //     };
        // }
        // );
        // this.props.handleAddOption(option); // call the props function here
        // e.target.elements.option.value = '';
        const error = this.props.handleAddOption(option);
        this.setState({
            error: error,
            // error   ES6 object shorthand
        });
        if (!error) {
            e.target.elements.option.value = '';
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleAddOption}>

                    <input type='text' name='option' />
                    <button >Add Option</button>

                </form>
            </div>
        );
    }
}

