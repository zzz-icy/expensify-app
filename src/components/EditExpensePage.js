import React from 'react';


const EditExpensePage = () => (
    <div>
        This is my EditExpensePage component!
    </div>
);
export default EditExpensePage;