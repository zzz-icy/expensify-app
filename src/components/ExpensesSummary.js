import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import SelectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export class ExpensesSummary extends React.Component {


    render() {
        const expenseWord = this.props.expensesCount === 1 ? 'expense' : 'expenses';
        const formattedExpneseTotal = numeral(this.props.expensesTotal / 100).format('$0,0.00');
        return (
            <div>
                <h2>
                    Viewing&nbsp;
                {this.props.expensesCount}&nbsp;
                {expenseWord} totalling&nbsp;
                {formattedExpneseTotal}&nbsp;
                </h2>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: SelectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
