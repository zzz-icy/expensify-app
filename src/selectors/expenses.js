// get visible expenses


const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter(
        (item) => {
            const startDateMatch = typeof startDate !== 'number' || item.createdAt >= startDate;
            const endDateMatch = typeof endDate !== 'number' || item.createdAt <= endDate;
            const textMatch = item.description.toLowerCase().includes(text.toLowerCase());

            // figure out if expenses.description as the text variable string inside of it
            // includes, The includes() method determines whether an array includes a certain element, returning true or false as appropriate.
            // convert both string to lower case

            // sort, in place and retur nthe array
            // sort(compareFunction(a,b)) 
            // return 1, b to lower index
            // return 0, unchanged
            // return -1, a to lower index
            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
            return startDateMatch && endDateMatch && textMatch;

        }
    ).sort(
        (a, b) => {
            if (sortBy === 'date') {
                return (a.createdAt < b.createdAt) ? 1 : -1;// latest date first
            }
            if (sortBy === 'amount') {
                return (a.amount < b.amount) ? 1 : -1; // higher amount first
            }
        }
    );
};

export default getVisibleExpenses;