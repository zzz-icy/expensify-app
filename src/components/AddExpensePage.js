import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'; // will get error TypeError: (0 , _someModule2.default) is not a function(â€¦)_

const AddExpensePage = (props) => (
    <div>
        <h1> This is my AddExpensePage component!</h1>
        <ExpenseForm
            onSubmit={(expense) => {
                // console.log(expense);
                props.dispatch(addExpense(expense));
                props.history.push('/');
            }}
        />
    </div>
);
export default connect()(AddExpensePage); // then we have dispatch as props