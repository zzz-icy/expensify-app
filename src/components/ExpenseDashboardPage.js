import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';



const ExpenseDashboardPage = () => (
    <div>
        This is my ExpenseDashboardPage component!
        <ExpensesSummary />

        <ExpenseListFilters />
        <ExpenseList />
    </div>
);


export default ExpenseDashboardPage;