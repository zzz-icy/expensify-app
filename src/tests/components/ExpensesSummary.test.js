import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import "../setupTests.js";
import expenses from '../fixtures/expenses';

test('should render correctly with 2 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={100} />)
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render correctly with 1 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={100} />)
    expect(toJSON(wrapper)).toMatchSnapshot();
});