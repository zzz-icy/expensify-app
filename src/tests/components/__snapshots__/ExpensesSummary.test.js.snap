// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with 1 expenses 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense
     totalling 
    $1.00
     
  </h2>
</div>
`;

exports[`should render correctly with 2 expenses 1`] = `
<div>
  <h2>
    Viewing 
    2
     
    expenses
     totalling 
    $1.00
     
  </h2>
</div>
`;
